### PR TITLE
🔨 (notebook) Create functions for historicization

### PR DESCRIPTION
**Pull request by loco-philippe**
## Purpose

Create examples (notebook `test_function_histo.md`) of functions to move historicization from a period to another and to return indicators values from historicization table.

## Proposal

- [x] function `to_df_histo_up` (calculate an historicization for a new period)
- [x] function `histo_up` (add new historicization with new period)
- [x] function `temporal_indicator` (extend indicators to historicized data)

### CODE REVIEW

**Comment by loco-philippe** [comment]
@jmaupetit 

- C'est principalement le fichier `test_function_histo.md` qui est à regarder
- La fonction `temporal_indicator` n'est pas encore terminée

**Comment by loco-philippe** [comment]
@jmaupetit 

- Le fichier `test_function_histo.md` est maintenant complet (fonction `temporal_indicator` terminée)

File: src/prefect/tests/conftest.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 @pytest.fixture(autouse=True, scope="session")
 def prefect_test_fixture():
     """Autouse the prefect test context."""
-    with prefect_test_harness():
+    with prefect_test_harness(60):
```
Comment (jmaupetit) [inline src/prefect/tests/conftest.py:16]: IMO should be committed separately.

File: src/prefect/indicators/models.py [hunk 2]
```diff
@@ -16,6 +16,14 @@ class Level(IntEnum):
...
     CITY = 4
 
+class LevelName(Enum):
+    """Administrative level names."""
+
+    NATIONAL = "national_group"
+    REGION = "region"
+    DEPARTMENT = "department"
+    EPCI = "epci"
+    CITY = "city"
```
Comment (jmaupetit) [inline src/prefect/indicators/models.py:26]: The alternative is to use the `Level` enum keys!

```python
assert Level.NATIONAL.name.lower() == "national"
```

File: src/notebook/indicators/config_indicators.json
Comment (jmaupetit) [inline src/notebook/indicators/config_indicators.json:1]: Not sure this is something we would use. IRL discussion concluded that it should be in the code, and we should be move indicators to classes that inherit from a `BaseIndicator` abstract class used as an interface.

File: src/notebook/misc/test_fonction_histo.md [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+             init_table: str ='histo', 
+             final_table: str ='histo', 
+             if_exists: str ='append') -> pd.DataFrame:
+    '''add new historicization with new period'''
+    
+    delta_period = { IndicatorPeriod.DAY: timedelta(days=1), 
+                     IndicatorPeriod.WEEK: timedelta(days=7), 
+                     IndicatorPeriod.MONTH: relativedelta(months=1), 
+                     IndicatorPeriod.QUARTER: relativedelta(months=3), 
+                     IndicatorPeriod.YEAR: relativedelta(year=1)}
```
[the remark this answers is not in the document]
Comment (loco-philippe) [reply src/notebook/misc/test_fonction_histo.md:154]: Done, i add this class in `src/prefect/indicators/models.py`

```python
class DurationPeriod(Enum):
    """duration of an IndicatorPeriod."""

    DAY = timedelta(days=1)
    WEEK = timedelta(days=7) 
    MONTH = relativedelta(months=1)
    QUARTER = relativedelta(months=3)
    YEAR = relativedelta(year=1)
```

File: src/notebook/indicators/config_indicators.json
Comment (loco-philippe) [reply src/notebook/indicators/config_indicators.json:1]: I agree, a class for each indicator is a good solution.

It remains to define which options we retain -> (see presentation added to the notebook) and if `BaseIndicator` is an abstract class or a parent class with several abstract methods.

File: src/prefect/tests/conftest.py (hunk 1 above)
Comment (loco-philippe) [reply src/prefect/tests/conftest.py:16]: done

File: src/prefect/indicators/models.py (hunk 2 above)
Comment (jmaupetit) [reply src/prefect/indicators/models.py:27]: You forgot to remove this Enum :wink:

File: src/prefect/indicators/models.py [hunk 4]
```diff
@@ -39,3 +48,18 @@ class Indicator(BaseModel):
...
     timestamp: datetime
+
+class InstancePeriod(BaseModel):
+    """period with a start time."""
+
+    period: IndicatorPeriod
+    start: datetime
+
+class DurationPeriod(Enum):
+    """duration of an IndicatorPeriod."""
```
[the remark this answers is not in the document]
Comment (loco-philippe) [reply src/prefect/indicators/models.py:59]: done

File: src/prefect/indicators/models.py [hunk 5]
```diff
@@ -39,3 +48,18 @@ class Indicator(BaseModel):
     value: float
     extras: Optional[dict] = None
     timestamp: datetime
+
+class InstancePeriod(BaseModel):
+    """period with a start time."""
```
[the remark this answers is not in the document]
Comment (loco-philippe) [reply src/prefect/indicators/models.py:53]: done

**Pull request merged**